Fix for 2PC causing WAL to grow too large

reidHoruff opened the pull request:
Consider the following single column family scenario:
prepare in log A
commit in log B
*WAL is too large, flush all CFs to releast log A*
*CFA is on log B so we do not see CFA is depending on log A so no flush is requested*

To fix this we must also consider the log containing the prepare section when determining what log a CF is dependent on.

siying commented:
I suggest we set a flag when this condition is not satisified: https://github.com/facebook/rocksdb/blob/5.0.fb/db/db_impl.cc#L869

If the flag is set, we can by pass the check in https://github.com/facebook/rocksdb/blob/5.0.fb/db/db_impl.cc#L4689 and trigger the extra CF flush if possible.

reidHoruff commented:
Siying, I believe I have addressed your concerns with the latest commit. I do not set the getting_flushed flag until all data in that log has been committed to a memtable and that memtable has been flush requested. It may take > 1 calls to FlushColumnFamilies() to release the oldest log when 2pc is enabled.

siying commented:
@reidHoruff would that cause FlushColumnFamilies() being called every time a Write() is called before it is ready to be dropped? That could be a huge performance regression.

reidHoruff commented:
@siying In most scenarios FlushColumnFamilies() should only be called once per oversized WAL. Most calls to FlushColumnFamilies() will flush all memtables necessary to release that log (setting flush_requested) the only scenario in which FlushColumnFamilies() will be called multiple times per oversized WAL is when there is a long outstanding prepare section, ie in an XA scenario. This should fix our current production issue without performance penalty.

reidHoruff commented:
@siying I will work on a better fix for the XA scenario but this should be good for our current bug.

siying commented:
@reidHoruff  "The only scenario in which FlushColumnFamilies() will be called multiple times per oversized WAL is when there is a long outstanding prepare section, ie in an XA scenario.". Why "long outstanding"?  The prepare phase doesn't need to be very long to trigger a recheck with no operation. Before all outstanding transactions are committed, all writes will trigger a recheck.

siying commented on db/db_impl.cc:
```diff
@@ -5019,19 +5019,19 @@ void DBImpl::FlushColumnFamilies() {
       continue;
     }
-    if (cfd->GetLogNumber() <= flush_column_family_if_log_file) {
+    if (cfd->OldestLogReferenced() <= oldest_alive_log) {
```
How do we make sure flushing this memtable will solve the problem? I mean there will not be pending transaction depending on the log to delete? Do we have the extra check that flushing the memtable will release the log file to delete?

siying commented on db/db_impl.cc:
```diff
@@ -5010,28 +5010,39 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
+  auto oldest_alive_log = alive_log_files_.begin()->number;
+  auto oldest_log_with_uncommited_prep = FindMinLogContainingOutstandingPrep();
+
+  if (unable_to_flush_oldest_log_ &&
```
If `unable_to_flush_oldest_log_ ` is false, can we return before calculating `oldest_alive_log` and `oldest_log_with_uncommited_prep`?

reidHoruff replied:
if unable_to_flush_oldest_log is false then we want to go forward with the flush so we need those values?

siying replied:
Then I'm totally confused.

siying commented on db/db_impl.cc:
```diff
@@ -5041,6 +5052,26 @@ void DBImpl::FlushColumnFamilies() {
+    } else {
+      Log(InfoLogLevel::WARN_LEVEL, immutable_db_options_.info_log,
+          "Unable to release oldest log due to uncommited transaction");
+      unable_to_flush_oldest_log_ = true;
```
In order to solve the polling check, I suggest a flag to skip the alive_log_files_.begin()->getting_flushed check. This should be set this in FindObsoleteFiles() when the flush has finished.

Now the polling will still  happen.

reidHoruff replied:
I don't understand this comment.

siying commented on db/column_family.cc:
```diff
@@ -492,6 +492,22 @@ ColumnFamilyOptions ColumnFamilyData::GetLatestCFOptions() const {
   return BuildColumnFamilyOptions(initial_cf_options_, mutable_cf_options_);
 }
 
+uint64_t ColumnFamilyData::OldestLogReferenced() {
```
Rename it to something like OldestLogToKeep() or something like that?

For non-2PC case, nothing is referenced, but it still returns a log number to keep.

siying commented on db/db_impl.cc:
```diff
@@ -5010,28 +5010,39 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
+    // the oldest alive log STILL contains uncommited transaction so there
+    // is still nothing that we can do.
+    return;
+  }
```
Can you guard it with 2PC? It feels safer not to expose this logic to all the users too fast.

siying commented on db/column_family.cc:
```diff
@@ -492,6 +492,22 @@ ColumnFamilyOptions ColumnFamilyData::GetLatestCFOptions() const {
 
+uint64_t ColumnFamilyData::OldestLogReferenced() {
+  auto current_log = GetLogNumber();
+  auto imm_prep_log = imm()->GetMinLogContainingPrepSection();
```
Similarly, I think it may be a better idea to still guard line 496 to 506 with if 2PC.
